### PR TITLE
2.1/create referee

### PR DESCRIPTION
**Pull request by koonwill**
Done please check

### CODE REVIEW

**Comment by Tezigudo** [comment]
LGTM but I will test it first

File: main.py [hunk 1]
```diff
@@ -78,7 +94,30 @@ async def add_data():
...
+    """
+    Receive username and password from body and check if user exists in database and password is correct
+    then set cookie with the access token that expires in 24 days, after that redirect to homepage
+    """
+    if await check_user(id_body) and await check_password(id_body):
+        access_token = create_access_token(id_body.username)
+        response = RedirectResponse(
+            url='/test/verify', status_code=302)  # change URL to homepage
+        response.set_cookie(key="access_token",
+                            value=f"Bearer {access_token}", httponly=True, expires=2073600)
```
Comment (Tezigudo) [inline main.py:111]: 24 day is too long, 30 min - 1 hr is enough

**Review by Tezigudo — APPROVED** [review]

File: models.py [hunk 2]
```diff
@@ -3,6 +3,12 @@
 from typing import List, Optional
 import datetime
 
+class Schedule(BaseModel):
+    pass
```
Comment (Tezigudo) [inline models.py:7]: remove it

File: requirements.txt [hunk 3]
```diff
@@ -1,3 +1,7 @@
 fastapi[all]
+motor
+beanie
```
Comment (Tezigudo) [inline requirements.txt:3]: remove it

File: router/auth.py [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+from models import RefereeIdBody, SportScheduleBody
+from utils import error_handler
+
+router = APIRouter(prefix='/auth',
+                   tags=["auth"],
+                   responses={ 404: {"description": "Not found"}})
+
+oauth2_scheme = OAuth2PasswordBearerWithCookie(tokenUrl="/auth/token")
+
+SECURE = config("SECURE", default=False, cast=bool)
```
Comment (Tezigudo) [inline router/auth.py:21]: name is not clear, change it pls

File: sample.env [hunk 5]
```diff
@@ -1 +1,5 @@
-MONGO_URL=YOUR-MONGO_URL-HERE
\ No newline at end of file
+MONGO_URL=YOUR-MONGO_URL-HERE
+JWT_SECRET = YOUR-SECRET-KEY
+JWT_ALGORITHM = YOUR-ALGORITHM
+COOKIE_NAME = YOUR-COOKIE-NAME
+SECURE = False # Change to True if prod, False if testing.
```
Comment (Tezigudo) [inline sample.env:5]: HTTP_COOKIE_SECURE? or COOKIE_SECURE?

**Review by Tezigudo — CHANGES_REQUESTED** [review]
change based on my comment

File: auth/cookie.py [hunk 6]
```diff
@@ -23,7 +23,7 @@ async def __call__(self, request: Request) -> Optional[str]:
         scheme, param = get_authorization_scheme_param(authorization)
         if not authorization or scheme.lower() != "bearer":
             if self.auto_error:
-              raise HTTPException(status_code=302, detail="Not authorized", headers = {"Location": "/login"} )
+              raise HTTPException(status_code=302, detail="Not authorized", headers = {"WWW-Authenticate": "Bearer", "Location": "https://www.google.com/"} )
```
[the remark this answers is not in the document]
Comment (koonwill) [reply auth/cookie.py:26]: Location is redirect to the page
In this case I set it to google if not authorize for testing

File: db.py [hunk 7]
```diff
@@ -6,3 +6,6 @@
 mongo_connection = client['referee']
 
 sport_schedule_connection = mongo_connection['SportSchedule']
+sport_connection = mongo_connection['Sport']
+sport_type_connection = mongo_connection['SportType']
```
[the remark this answers is not in the document]
Comment (koonwill) [reply db.py:10]: for further implementation I have used sport_connection to easily get all of the sport on /api/schedule/sport that was merge before, So we didn't need loop or slicing it

[the remark this answers is not in the document]
Comment (Tezigudo) [reply db.py:10]: what about sport_type?

File: router/auth.py [hunk 8]
```diff
@@ -0,0 +1,38 @@
...
+import sys
+from pathlib import Path
+
+file = Path(__file__).resolve()
+package_root_directory = file.parents[1]
+sys.path.append(str(package_root_directory))
+from models import RefereeIdBody, SportScheduleBody
+from utils import error_handler
+
+router = APIRouter(prefix='/auth',
```
[the remark this answers is not in the document]
Comment (Tezigudo) [reply router/auth.py]: wtf where/api

File: router/auth.py [hunk 9]
```diff
@@ -0,0 +1,45 @@
...
+                   tags=["auth"],
+                   responses={ 404: {"description": "Not found"}})
+
+oauth2_scheme = OAuth2PasswordBearerWithCookie(tokenUrl="api/auth/token")
+
+COOKIE_SECURE = config("SECURE", default=False, cast=bool)
+
+@error_handler
+@router.post('/token', status_code=201)
+async def login(response: Response, form_data:OAuth2PasswordRequestForm = Depends()):
```
Comment (Tezigudo) [inline router/auth.py:25]: why it **asyncronous** now? we didnt need `async` and `await` now because we use pymongo

File: router/auth.py [hunk 10]
```diff
@@ -0,0 +1,45 @@
...
+    if await check_user(form_data) and await check_password(form_data):
+        access_token = create_access_token(form_data.username)
+        response.set_cookie(key="access_token",
+                            value=f"Bearer {access_token}", httponly=True, expires=3600, secure=COOKIE_SECURE)
+        return {"message": "Login successful"}
+    raise HTTPException(401, "Invalid credentials")
+
+@error_handler
+@router.get('/users')
+async def get_user(user = Depends(get_current_user)):
```
Comment (Tezigudo) [inline router/auth.py:39]: this too

File: router/auth.py [hunk 11]
```diff
@@ -0,0 +1,45 @@
...
+    return user
+
+@router.get('/test', dependencies=[Depends(oauth2_scheme)])
+async def toast():
```
Comment (Tezigudo) [inline router/auth.py:43]: this too

File: auth/cookie.py [hunk 12]
```diff
@@ -0,0 +1,29 @@
...
+        scheme_name: Optional[str] = None,
+        scopes: Optional[Dict[str, str]] = None,
+        auto_error: bool = True,
+    ):
+        if not scopes:
+            scopes = {}
+        flows = OAuthFlowsModel(password={"tokenUrl": tokenUrl, "scopes": scopes})
+        super().__init__(flows=flows, scheme_name=scheme_name, auto_error=auto_error)
+
+    async def __call__(self, request: Request) -> Optional[str]:
```
Comment (Tezigudo) [inline auth/cookie.py:20]: oh I see it now its ok

**Review by Tezigudo — APPROVED** [review]
resolve conversation first

**Pull request merged**